gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/DomainPasswordPoliciesUpgrader.java

```diff
@@ -0,0 +1,113 @@
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gravitee.am.management.service.impl.upgrades;
+
+
+import io.gravitee.am.model.ReferenceType;
+import io.gravitee.am.model.SystemTask;
+import io.gravitee.am.model.SystemTaskStatus;
+import io.gravitee.am.service.DomainService;
+import io.gravitee.am.service.PasswordPolicyService;
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.core.SingleSource;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+
+import static io.gravitee.am.management.service.impl.upgrades.UpgraderOrder.DOMAIN_PASSWORD_POLICIES_UPGRADER;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Component
+@Slf4j
+public class DomainPasswordPoliciesUpgrader extends SystemTaskUpgrader {
+
+    private static final String TASK_ID = "domain_password_settings_migration";
+    private static final String UPGRADE_NOT_SUCCESSFUL_ERROR_MESSAGE =
+            "Domain Password Policy can't be upgraded, other instance may process them or an upgrader has failed previously";
+    public static final String PASSWORD_POLICY_NAME_DEFAULT = "default";
+
+    @Autowired
+    private DomainService domainService;
+
+    @Autowired
+    private PasswordPolicyService passwordPolicyService;
+
+    @Override
+    protected Single<Boolean> processUpgrade(String instanceOperationId, SystemTask task, String previousOperationId) {
+        return updateSystemTask(task, (SystemTaskStatus.ONGOING), previousOperationId)
+                .flatMap(updatedTask -> {
+                    if (updatedTask.getOperationId().equals(instanceOperationId)) {
+                        return migrateDomainPasswordSettings(updatedTask);
+                    } else {
+                        return Single.error(new IllegalStateException("Task " + getTaskId() + " already processed by another instance : trigger a retry"));
+                    }
+                });
+    }
+
+    private SingleSource<? extends Boolean> migrateDomainPasswordSettings(SystemTask task) {
+        return domainService.findAll()
+                .flatMapCompletable(domains ->
+                        Flowable.fromIterable(domains)
+                                .flatMapCompletable(domain -> {
+                                    if (domain.getPasswordSettings() != null) {
```

Comment:
what do you think about using Single.merge?

The type is Single and it can be switched to Completable only at the end like this:

  ```
 private Completable migrateDomainPasswordSettings(SystemTask task) {
        return domainService.findAll()
                .flatMap(domains -> {
                    var list = new ArrayList<Single<PasswordPolicy>>();
                    for(var domain : domains) {
                        if (domain.getPasswordSettings() != null) {
                            final var policy = domain.getPasswordSettings().toPasswordPolicy();
                            policy.setName(PASSWORD_POLICY_NAME_DEFAULT);
                            policy.setDefaultPolicy(true);
                            policy.setReferenceId(domain.getId());
                            policy.setReferenceType(ReferenceType.DOMAIN);
                            Date now = new Date();
                            policy.setCreatedAt(now);
                            policy.setUpdatedAt(now);
                            list.add(passwordPolicyService.create(policy, null));
                        }
                    }
                    return Single.merge(list).toList();
                })
                .doOnError(err -> updateSystemTask(task, (SystemTaskStatus.FAILURE), task.getOperationId()).subscribe())
                .flatMapCompletable(results -> updateSystemTask(task, SystemTaskStatus.SUCCESS, task.getOperationId()).ignoreElement());
    }
```

What do you think?

Reply:
I added a new method at domainService to return a Flowable instead.